third_party/txt/src/txt/paragraph_txt.cc

```diff
@@ -1732,11 +1738,11 @@ std::vector<Paragraph::TextBox> ParagraphTxt::GetRectsForRange(
     if (line_box_metrics.find(line_number) == line_box_metrics.end()) {
       if (line.end_index != line.end_including_newline &&
           line.end_index >= start && line.end_including_newline <= end) {
-        SkScalar x = line_widths_[line_number];
-        // Move empty box to center if center aligned and is an empty line.
-        if (x == 0 && !isinf(width_) &&
-            paragraph_style_.effective_align() == TextAlign::center) {
-          x = width_ / 2;
+        SkScalar x;
+        if (glyph_lines_[line_number].positions.empty()) {
```

Comment:
Use `newline_box_positions.find(line_number)` here

Reply:
If newline is in the selected range, it's position is evaluated in one related bidi run in the previous loop.
So we are sure about the value.
I also don't know how to handle `find()` here or what to do if it's value becomes `end()`

Reply:
Try this:
```
        auto it = newline_x_positions.find(line_number);
        if (it != newline_x_positions.end()) {
          x = it->second;
        } else {
          x = GetLineXOffset(0, line_number, false);
        }
```

If there are glyphs on the line, then `newline_x_positions` will have a value for the line and that value should be used.  If not, then use the offset for an empty line.

Reply:
You already handled that case in the if statement.
```
        if (***glyph_lines_[line_number].positions.empty()***) {
          x = GetLineXOffset(0, line_number, false);
        } else {
          x = newline_x_positions[line_number];
        }
```

Reply:
The goal is to ensure that the code does not accidentally use a `newline_x_positions` value that was never set.

If something goes wrong and  `glyph_lines_[line_number].positions` is nonempty but `newline_x_positions[line_number]` was never set, then `newline_x_positions[line_number]` is going to yield a default value.  That will produce an unintended result.

Checking `newline_x_positions.find(line_number)` will ensure that this does not happen.

Reply:
Done.

Reply:
The `newline_x_positions.find(line_number)` check should replace the `glyph_lines_[line_number].positions.empty()` check.  Both checks should always have the same value if everything is working as expected, but using `newline_x_positions` is safer as discussed above.

Remove the `if glyph_lines_[line_number].positions.empty()` block, and just keep the block inside the `else` below.

Reply:
I got it now.
Done.
Thanks for your review